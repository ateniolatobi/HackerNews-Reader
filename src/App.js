import React, { Component } from "react";
import logo from "./logo.svg";
import "./App.css";
import axios from "axios";
import { sortBy } from "lodash";
import PropTypes from "prop-types";
import classnames from "classnames";
const DEFAULT_QUERY = "redux";
const PATH_BASE = "https://hn.algolia.com/api/v1";
const PATH_SEARCH = "/search";
const PARAM_SEARCH = "query=";
const PARAM_PAGE = "page=";

const largeColumn = {
  width: "40%"
};
const midColumn = {
  width: "30%"
};
const smallColumn = {
  width: "10%"
};

const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, "title"),
  AUTHOR: list => sortBy(list, "author"),
  COMMENTS: list => sortBy(list, "num_comments").reverse(),
  POINTS: list => sortBy(list, "points").reverse()
};

const updateSearchTopStoryState = result => prevState => {
  const { hits, page } = result;
  const { list, key } = prevState;
  const oldHit = list && list[key] ? list[key].hits : [];

  const updateHits = [...oldHit, ...hits];
  console.log("Search is ", key);
  return {
    list: { ...list, [key]: { hits: updateHits, page } },
    isLoading: false
  };
};
// console.log("not in shit " + SORTS["NONE"]);

const isSearched = searchTerm => item =>
  item.title.toLowerCase().includes(searchTerm.toLowerCase());

class Search extends Component {
  componentDidMount() {
    if (this.input) {
      this.input.focus();
    }
  }

  render() {
    const { search, searchText, children, onSubmit } = this.props;

    return (
      <form onSubmit={onSubmit}>
        {" "}
        {children}{" "}
        <input
          type="text"
          value={search}
          onChange={searchText}
          onSubmit={event => event.preventDefault()}
          ref={node => {
            this.input = node;
          }}
        />{" "}
        <button> Submit</button>
      </form>
    );
  }
}

const Loading = () => <div>Loading...</div>;

const SORT = ({ onSort, children, sortKey, activeSortKey }) => {
  console.log("SORT sortkey is ", sortKey);
  const sortClass = classnames("button-inline", {
    "button-active": activeSortKey === sortKey
  });
  // console.log("sortCLass is ", sortClass.join(" "));
  return (
    <Button className={sortClass} onClick={() => onSort(sortKey)}>
      {children}
    </Button>
  );
};
class BookList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      sortkey: "NONE",
      isSortReverse: false
    };
    this.onSort = this.onSort.bind(this);
  }

  onSort(key) {
    // console.log("ONsoRT SORT KEY IS ", sortkey);
    // console.log("sortkey is ", this.state.sortkey);
    const { sortkey, isSortReverse } = this.state;
    const Reverse = sortkey === key && !isSortReverse;
    this.setState({ sortkey: key, isSortReverse: Reverse });
  }
  render() {
    const { list, onDismiss } = this.props;
    const { sortkey, isSortReverse } = this.state;
    const new_list = SORTS[sortkey](list);
    // console.log("is sortreverse is " + isSortReverse);
    const reverseList = isSortReverse ? new_list.reverse() : new_list;
    // console.log("onsort here is ", this.onSort);
    console.log("THe active sortkey is ", sortkey);
    return (
      <div className="table">
        <div className="sort-options">
          <SORT onSort={this.onSort} sortKey="NONE" activeSortKey={sortkey}>
            Originial Form
          </SORT>
          <SORT onSort={this.onSort} sortKey="TITLE" activeSortKey={sortkey}>
            Sort by Title
          </SORT>
          <SORT onSort={this.onSort} sortKey="AUTHOR" activeSortKey={sortkey}>
            Sort by author
          </SORT>

          <SORT sortKey="COMMENTS" onSort={this.onSort} activeSortKey={sortkey}>
            Sort by comments
          </SORT>
          <SORT sortKey="POINTS" onSort={this.onSort} activeSortKey={sortkey}>
            Sort by points
          </SORT>
        </div>
        {reverseList.map(item => {
          return (
            <div key={item.objectID} className="table-row">
              <span style={largeColumn}>
                <a href={item.url}>{item.title}</a>
              </span>
              <span style={midColumn}>{item.author}</span>
              <span style={smallColumn}>{item.date}</span>
              <span>
                <Button onClick={() => onDismiss(item.objectID)}>
                  Dismiss
                </Button>
              </span>
            </div>
          );
        })}
      </div>
    );
  }
}

const Button = ({ onClick, children, className }) => {
  return (
    <button onClick={onClick} className={className}>
      {children}
    </button>
  );
};

Button.propTypes = {
  onClick: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
  className: PropTypes.string
};

const withLoading = Component => ({ isLoading, ...rest }) =>
  isLoading ? <Loading /> : <Component {...rest} />;

const ButtonWithLoading = withLoading(Button);

Button.defaultProps = {
  className: ""
};

BookList.propTypes = {
  list: PropTypes.array.isRequired,
  onDismiss: PropTypes.func.isRequired
};

class App extends Component {
  constructor(props) {
    super(props);
    const search = DEFAULT_QUERY;
    const list = null;
    this.state = {
      list,
      search,
      key: "",
      error: null,
      willUnmount: false,
      isLoading: false,
      sortKey: "NONE",
      isSortReverse: false
    };
    this.searchText = this.searchText.bind(this);
    this.setsearchTopStories = this.setsearchTopStories.bind(this);
    this.needsToSearchTopResults = this.needsToSearchTopResults.bind(this);
    this.fetchTopStories = this.fetchTopStories.bind(this);
    this.dismiss = this.dismiss.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  // onSort(sortKey) {
  //   const isSortReverse = this.state.sortKey === sortKey && !this.isSortReverse;
  //   this.setState({ sortKey, isSortReverse });
  // }

  searchText(event) {
    this.setState({ search: event.target.value });
  }

  needsToSearchTopResults(key) {
    const { list } = this.state;
    return !list[key];
  }

  dismiss(i) {
    alert(i);
    const { key, list } = this.state;
    const { hits, page } = this.state.list[key];
    const new_hits = hits.filter(item => {
      return item.objectID !== i;
    });
    this.setState({
      list: { ...list, [key]: { hits: new_hits, page } }
    });
  }

  render() {
    const { search, list, key, error, isLoading } = this.state;
    const page = (list && list[key] && list[key].page) || 0;
    const hit_list = (list && list[key] && list[key].hits) || [];
    return (
      <div className="page">
        <div className="interactions">
          <Search
            search={search}
            searchText={event => {
              this.searchText(event);
            }}
            onSubmit={this.onSubmit}
          >
            Type to Search
          </Search>
        </div>
        {!error ? (
          <BookList list={hit_list} onDismiss={id => this.dismiss(id)} />
        ) : (
          <div className="interactions">
            <p>Something went wrong.</p>
          </div>
        )}
        <div className="interactions">
          <ButtonWithLoading
            onClick={() => {
              this.fetchTopStories(this.state.key, page + 1);
            }}
            isLoading={isLoading}
          >
            More...
          </ButtonWithLoading>
        </div>
      </div>
    );
  }

  onSubmit(event) {
    const { search, key } = this.state;
    this.setState({ key: search });
    if (this.needsToSearchTopResults(search)) {
      console.log("Triggered");
      this.fetchTopStories(search);
    }
    event.preventDefault();
  }

  fetchTopStories(searchTerm, page = 0) {
    // console.log("search term is ", searchTerm);
    // alert("Awio");
    this.setState({ isLoading: true });
    axios(
      `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}`
    )
      .then(
        result => !this.willUnmount && this.setsearchTopStories(result.data)
      )
      .catch(error => !this.willUnmount && this.setState({ error }));
  }

  setsearchTopStories(result) {
    this.setState(updateSearchTopStoryState(result));
    // this.setState(prevState => {
    //   const { hits, page } = result;
    //   const { list, key } = prevState;
    //   const oldHit = list && list[key] ? list[key].hits : [];

    //   const updateHits = [...oldHit, ...hits];
    //   console.log("Search is ", key);
    //   return {
    //     list: { ...list, [key]: { hits: updateHits, page } },
    //     isLoading: false
    //   };
    // });
  }

  componentDidMount() {
    this.willUnmount = false;
    const { search } = this.state;
    this.setState({ key: search });
    this.fetchTopStories(search);
  }

  componentWillUnmount() {
    this.setState({ willUnmount: true });
  }
}

export default App;

export { Search, BookList, Button };
