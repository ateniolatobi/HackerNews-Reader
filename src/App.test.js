import React from "react";
import { render } from "@testing-library/react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import App, { Search, Button, BookList } from "./App";
import { exportAllDeclaration } from "@babel/types";

Enzyme.configure({ adapter: new Adapter() });
describe("App", () => {
  it("Renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Search", () => {
  it("Renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Search>Search</Search>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(<Search>Type to Search</Search>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Button", () => {
  it("Renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Button>Give me more</Button>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(<Button>Give me more</Button>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("BookList", () => {
  const props = {
    list: [
      { title: "1", author: "1", num_comments: 1, points: 2, objectID: "y" },
      { title: "2", author: "2", num_comments: 1, points: 2, objectID: "z" }
    ]
  };
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<BookList {...props} />, div);
  });
  test("has a valid snapshot", () => {
    const component = renderer.create(<BookList {...props} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows two elements in list", () => {
    const element = shallow(<BookList {...props} />);
    expect(element.find(".table-row").length).toBe(2);
  });
});
