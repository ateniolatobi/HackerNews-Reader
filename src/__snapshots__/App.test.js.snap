// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
       
      Type to Search
       
      <input
        onChange={[Function]}
        onSubmit={[Function]}
        type="text"
        value="redux"
      />
       
      <button>
         Submit
      </button>
    </form>
  </div>
  <div
    className="table"
  />
  <div
    className="interactions"
  >
    <button
      onClick={[Function]}
    >
      More..
    </button>
  </div>
</div>
`;

exports[`App has a valid snapshot 2`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
       
      Type to Search
       
      <input
        onChange={[Function]}
        onSubmit={[Function]}
        type="text"
        value="redux"
      />
       
      <button>
         Submit
      </button>
    </form>
  </div>
  <div
    className="table"
  />
  <div
    className="interactions"
  >
    <button
      onClick={[Function]}
    >
      More..
    </button>
  </div>
</div>
`;

exports[`B has a valid snapshot 1`] = `
<form>
   
  Type to Search
   
  <input
    onSubmit={[Function]}
    type="text"
  />
   
  <button>
     Submit
  </button>
</form>
`;

exports[`BookList has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    />
    <span>
      <button
        onClick={[Function]}
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    />
    <span>
      <button
        onClick={[Function]}
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<form>
   
  Type to Search
   
  <input
    onSubmit={[Function]}
    type="text"
  />
   
  <button>
     Submit
  </button>
</form>
`;

exports[`S has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
       
      Type to Search
       
      <input
        onChange={[Function]}
        onSubmit={[Function]}
        type="text"
        value="redux"
      />
       
      <button>
         Submit
      </button>
    </form>
  </div>
  <div
    className="table"
  />
  <div
    className="interactions"
  >
    <button
      onClick={[Function]}
    >
      More..
    </button>
  </div>
</div>
`;

exports[`Search has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
       
      Type to Search
       
      <input
        onChange={[Function]}
        onSubmit={[Function]}
        type="text"
        value="redux"
      />
       
      <button>
         Submit
      </button>
    </form>
  </div>
  <div
    className="table"
  />
  <div
    className="interactions"
  >
    <button
      onClick={[Function]}
    >
      More..
    </button>
  </div>
</div>
`;

exports[`Search has a valid snapshot 2`] = `
<form>
   
   Search
   
  <input
    onSubmit={[Function]}
    type="text"
  />
   
  <button>
     Submit
  </button>
</form>
`;

exports[`has a valid snapshot 1`] = `
<form>
   
  Type to Search
   
  <input
    onSubmit={[Function]}
    type="text"
  />
   
  <button>
     Submit
  </button>
</form>
`;
